### DISCUSSION TITLE
Partial time partition mappings

### DISCUSSION THREAD

**Discussion by abkfenris** [dagster-io/dagster#16036] [Q&A] [answered]
I'm trying to set up some sensor data aggregation with assets and auto-materialization. The API that we're using prefers to be queried daily, but then we'd like to aggregate it to monthly (then eventually yearly or everything into a single asset).

```py
@asset(
    partitions_def=DailyPartitionsDefinition(start_date="2023-06-02", end_offset=1)
)
def daily_data():
    ...

@asset(
    deps=[daily_data],
    partitions_def=MonthlyPartitionsDefinition(start_date="2023-06-02", end_offset=1),
    auto_materialize_policy=AutoMaterializePolicy.eager()
)
def aggregate_monthly():
    ...
```

If I trigger a backfill of all `daily_data` for July, then `aggregate_monthly` will run, but it doesn't run for August data when backfilled or individual days are run. Similarly if I backfill June because the start date doesn't include the first of the month, `aggregate_monthly` never ends up running.

From what I can see, this is expected [by the default time partition mappings](https://docs.dagster.io/concepts/partitions-schedules-sensors/partitions#partition-dependencies), but is there a built in way to allow partitions to be auto materialized when only some of their upstream dependencies partitions exist (without writing a full new mapping)?

**Comment by sryza** [comment]
> but it doesn't run for August data when backfilled or individual days are run.

This is because the partition for August doesn't exist yet in `aggregate_monthly`, right? If you set `end_offset=1` on the `MonthlyPartitionsDefinition`, then the partition for August will exist as soon as August begins.

**Reply by abkfenris** [reply]
Whoops, the August partition does exist as I do have `end_offset=1`, but still does not update.

I just copied it wrong as I first read the docs to say that the offset needed to be negative to include the current, then futzed around till the partition appeared, then the argument was just off the screen as I was writing a simplified version here.

**Comment by sryza** [comment, marked as answer]
I believe you can address this by using a `TimeWindowPartitionMapping` with `allow_nonexistent_upstream_partitions=True".

```python
@asset(
    ins={"daily_data": AssetIn(
        partition_mapping=TimeWindowPartitionMapping(allow_nonexistent_upstream_partitions=True)
    )}
    partitions_def=MonthlyPartitionsDefinition(start_date="2023-06-02", end_offset=1),
    auto_materialize_policy=AutoMaterializePolicy.eager()
)
def aggregate_monthly():
    ...
```

**Reply by abkfenris** [reply]
Ah, ok. I tinkered with `TimeWindowPartitionMapping(allow_nonexistent_upstream_partitions=True)` but couldn't get it to work with the `deps=[]` kwarg, and the [docs](https://docs.dagster.io/_apidocs/partitions#dagster.TimeWindowPartitionMapping.allow_nonexistent_upstream_partitions) are a little vague about the exact impacts of `allow_nonexistent_upstream_partitions`, so I didn't spend too much time tinkering with it.

I had to switch to including `daily_data` in the `aggregate_monthly` function arguments, even though it's not using what's returned from `daily_data`.

```py
@asset(
    ins={"daily_data": AssetIn(
        partition_mapping=TimeWindowPartitionMapping(allow_nonexistent_upstream_partitions=True)
    )}
    partitions_def=MonthlyPartitionsDefinition(start_date="2023-06-02", end_offset=1),
    auto_materialize_policy=AutoMaterializePolicy.eager()
)
def aggregate_monthly(daily_data):
    ...
```

Thank you!

**Reply by sryza** [reply]
Ah if you set `dagster_type=Nothing` on the `AssetIn` then you won't need to include it in the function arguments. We're currently working on a more ergonomic way to do this.
